02_commands.html

```diff
@@ -1,206 +1,524 @@
+
+        <!-- TODO: really need a better way of constructing JSON literals -->
+      <li><p><a>Set the properties</a> of <var>object</var> with values ("error", <var>name</var>), ("message", <var>message</var>).</p></li>
+
+      <li><a>Send a response</a> with status <var>status</var> and data <var>data</var></li>
+
+    </ol>
+
+  </section>
+
+  <section>
+    <h3>Routing Requests</h3>
+
+    <!-- TODO: define what a command is -->
+
+    <p>Request routing is the process of going from a HTTP request to the series of steps needed to implement the <a>command</a> represented by that request.</p>
+
+    <p>A <a>remote end</a> may have a <dfn>url prefix</dfn>, which is a base url that is appended to the start of all WebDriver URLs. By default this is the empty string.</p>
+
+    <aside class="note">
+      <p>For example a remote end wishing to run alongside other services on <code>example.com</code> might set its <a>url prefix</a> to <code>/wd</code> so that a <a>new session</a> command would be invoked by sending a POST request to <code>/wd/session</code> rather than <code>/session</code>.</p>
+    </aside>
+
+    <p>In order to <dfn>match a request</dfn> given a <var>method</var> and <var>url</var>, the following steps must be taken:</p>
+
+    <ol>
+      <li><p>Let <var>endpoints</var> be a list containing each row in the <a>table of endpoints</a>.</p></li>
+
+   <li><p>Remove each entry in <var>endpoints</var> for which the concatenation of the <a>url prefix</a> and the production in the "url" column does not match <a>url</a> according to the rules in [[!URI-Template]].</p></li>
+
+      <li><p>If there are no entries in <var>endpoints</var>, return an <a>error</a> with code <a>unknown command</a></p></li>
+
+      <li><p>Remove each entry in <var>endpoints</var> for which the "method" column is not an exact case-sensitive match for <var>method</var>.</p></li>
+
+      <li><p>If there are no entries in <var>endpoints</var>, return an <a>error</a> with code <a>unknown method</a>.</p></li>
+
+      <li><p>There is now exactly one entry in <var>endpoints</var>; let <var>entry</var> be this entry.</p></li>
+
+
+      <!-- TODO: Check that URI-templates is being used in the right way here -->
+
+      <li><p>If the match in step 2 populated a template variable called "sessionId", let <var>session id</var> be the value it was populated with. Otherwise let <var>session id</var> be null.</p></li>
+
+      <li><p>If the match in step 2 populated a template variable called "elementId", let <var>element id</var> be the value it was populated with. Otherwise let <var>element id</var> be null.</p></li>
+
+      <li><p>Let <var>command</var> by the command represented by the "command" column of <var>entry</var></p></li>
+
+      <li><p>Return <a>success</a> with data <var>session id</var>, <var>element id</var> and <var>command</var></li>
```

Comment:
Right, so I'm not clear on how to say what the return type is, exactly. It's a tuple of (session id, element id, command) wrapped in "Success". "command" is a reference to the command being run, so it's like a function pointer or something rather than a type containing a pure value.